games/chapter5/examples/OOP_game.py

```diff
@@ -0,0 +1,424 @@
+        if direction == "up":
+            self.direction[1] -= 1
+        if direction == "down":
+            self.direction[1] += 1
+        if direction == "left":
+            self.direction[0] -= 1
+        if direction == "right":
+            self.direction[0] += 1
+
+    def unset_path(self, direction: str) -> None:
+        if direction == "up":
+            self.direction[1] += 1
+        if direction == "down":
+            self.direction[1] -= 1
+        if direction == "left":
+            self.direction[0] += 1
+        if direction == "right":
+            self.direction[0] -= 1
+
+
+class App:
+    def __init__(
+        self, flags=RESIZABLE, width=960, height=540, title="My Game"
+    ):
+        pygame.init()
+        self.size = [width, height]
+        self.screen = pygame.display.set_mode(self.size, flags)
+        pygame.display.set_caption(title, title)
+
+        self.running = True
+
+        self.create_objects()
+
+    def create_objects(self):
+        pass
+
+    def check_events(self, event):
+        pass
+
+    def update_display(self):
+        pass
+
+    def move_objects(self):
+        pass
+
+    def check_collisions(self):
+        pass
+
+    def mainloop(self):
+        while self.running:
+            for event in pygame.event.get():
+                if event.type == QUIT:
+                    self.running = False
+                    break
+                else:
+                    self.check_events(
+                        event
+                    )  # this will handle checking for user input
+                    # such as KEYUP and MOUSEBUTTONDOWN events needed to run the game
+            self.check_collisions()  # checks collisions between bullet/tank and targets
+            self.move_objects()  # moves each object on the screen
+            self.update_display()  # redraws updated objects onto the screen
+            pygame.display.update()  # pygame’s method to show the updated screen
+            time.sleep(0.01)  # not necessary; it's a frame cap
+        pygame.quit()
+
+
+class Tank_Game(App):
+    def __init__(self):
+        super().__init__(title="Tanks")
+
+        self.playerscore = 0  # the player's score
+
+        # this can be changed, it's the number of targets allowed at a time.
+        self.NUM_TARGETS = 3
+
+        # sets the display icon to the TankIcon.png provided
+        pygame.display.set_icon(pygame.image.load("./TankIcon.png"))
+
+    def create_objects(self):
+        """
+        This creates the initial objects seen when the game
+        first starts up.
+        """
+        # tank
+        self.tank = Tank(speed=TANKSPEED)
+        self.tank.moveto(
+            (
+                self.size[0] / 2 - self.tank.size[0],  # move to middle x
+                self.size[1] - self.tank.size[1],  # move to bottom y
+            )
+        )
+
+        # targets
+        self.targets = [Target(speed=[0, 0]) for i in range(self.NUM_TARGETS)]

```

Comment:
AttributeError: 'Tank_Game' object has no attribute 'NUM_TARGETS', though I'm not sure why.

Reply:
Just realized. It's probably because the "self.NUM_TARGETS comes after the super().__init__ and super().__init__ is what calls create_objects. I'll change that rn.